### ISSUE TITLE
Windows 7 gyp build error 

### ISSUE THREAD

**Issue by moshest** [cryptocoinjs/secp256k1-node#20]
Error log:

```
node "C:\Program Files\nodist\bin\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild
'.' is not recognized as an internal or external command, operable program or batch file.
gyp: Call to './utils/has_lib.sh gmpxx' returned exit status 1. while trying to load binding.gyp
```

**Comment by wanderer** [comment]
hi @moshest i don't have a windows machine to test on at the moment. Ill check back when I do.

**Comment by moshest** [comment]
OK. 

Just for reference: tested on node `v0.10` and `v0.12`.

Full error log from node `v0.12`:

```
'.' is not recognized as an internal or external command,
operable program or batch file.
gyp: Call to './utils/has_lib.sh gmpxx' returned exit status 1. while trying to
load binding.gyp
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (C:\Program Files\nodist\bin\node_mo
dules\npm\node_modules\node-gyp\lib\configure.js:340:16)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1074
:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "node" "C:\\Program Files\\nodist\\bin\\node_modules\\npm\\node
_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\owner\git\nodejs\Anternet\prototype\node_modules\secp256k1

gyp ERR! node -v v0.12.4
gyp ERR! node-gyp -v v0.13.0
gyp ERR! not ok
npm ERR! secp256k1@0.0.15 install: `node-gyp rebuild`
npm ERR! Exit status 1
```

**Comment by wanderer** [comment]
well ./utils/has_lib.sh is a bash script. I'm would assume widows doesn't
have bash.

**Comment by wanderer** [comment]
so the bash is just to detect whether to complile using gmp or not. If your on windows you probably won't have gmp either.  What happens if you just set [this](https://github.com/wanderer/secp256k1-node/blob/master/binding.gyp#L5) to false?

**Comment by moshest** [comment]
Got other error:

    Building the projects in this solution one at a time. To enable parallel build,
    please add the "/m" switch.
    
       secp256k1.cc
    ..\secp256k1.cc(5): fatal error C1083: Cannot open include file: './secp256k1-s
    rc/include/secp256k1.h': No such file or directory [E:\Users\owner\Downloads\se
    cp256k1-node\build\secp256k1.vcxproj]
    gyp ERR! build error
    gyp ERR! stack Error: `C:\Windows\Microsoft.NET\Framework\v4.0.30319\msbuild.exe
    ` failed with exit code: 1

**Comment by moshest** [comment]
Now there is a log of errors like:

```
          E:\Program Files (x86)\Visual Studio 2012\VC\include\stdint.h(14) : s
  ee declaration of 'uint32_t'
c:\users\owner\downloads\secp256k1-node\secp256k1-src\src\field_10x26_impl.h(23
3): error C2146: syntax error : missing ';' before identifier 'z0' [C:\Users\ow
ner\Downloads\secp256k1-node\build\secp256k1.vcxproj]
c:\users\owner\downloads\secp256k1-node\secp256k1-src\src\field_10x26_impl.h(23
3): error C2065: 'z0' : undeclared identifier [C:\Users\owner\Downloads\secp256
k1-node\build\secp256k1.vcxproj]
c:\users\owner\downloads\secp256k1-node\secp256k1-src\src\field_10x26_impl.h(23
3): error C2065: 'z1' : undeclared identifier [C:\Users\owner\Downloads\secp256
k1-node\build\secp256k1.vcxproj]
c:\users\owner\downloads\secp256k1-node\secp256k1-src\src\field_10x26_impl.h(23
6): error C2065: 'z0' : undeclared identifier [C:\Users\owner\Downloads\secp256
k1-node\build\secp256k1.vcxproj]
c:\users\owner\downloads\secp256k1-node\secp256k1-src\src\field_10x26_impl.h(23
6): error C2065: 'z1' : undeclared identifier [C:\Users\owner\Downloads\secp256
k1-node\build\secp256k1.vcxproj]
c:\users\owner\downloads\secp256k1-node\secp256k1-src\src\field_10x26_impl.h(23
9): error C2275: 'uint32_t' : illegal use of this type as an expression [C:\Use
rs\owner\Downloads\secp256k1-node\build\secp256k1.vcxproj]
```

and more...

**Comment by wanderer** [comment]
If anyone has a window machine and knows how to compile https://github.com/bitcoin/secp256k1 on it please I would love to know/ have help

**Comment by simkuns** [comment]
Installed MSVS 2015 Express and Python 2.7

Added https://github.com/bitcoin/secp256k1 files to "secp256k1-src" directory

Edited "package.json" line 8-9 to:

``` c
    "install": "node-gyp rebuild --msvs_version=2015",
    "rebuild": "node-gyp rebuild --msvs_version=2015"
```

Edited "bindings.gyp" line 5-47 to:

``` ini
      "with_gmp%": "false"
    },
    "sources": [
      "functions.cc",
      "util.cc",
      "secp256k1.cc",
      "secp256k1-src/src/secp256k1.c",
      "secp256k1-src/src/ecdsa.h",
      "secp256k1-src/src/ecdsa_impl.h",
      "secp256k1-src/src/eckey.h",
      "secp256k1-src/src/eckey_impl.h",
      "secp256k1-src/src/ecmult_const.h",
      "secp256k1-src/src/ecmult_const_impl.h",
      "secp256k1-src/src/ecmult_gen.h",
      "secp256k1-src/src/ecmult_gen_impl.h",
      "secp256k1-src/src/ecmult.h",
      "secp256k1-src/src/ecmult_impl.h",
      "secp256k1-src/src/field_10x26.h",
      "secp256k1-src/src/field_10x26_impl.h",
      "secp256k1-src/src/field_5x52.h",
      "secp256k1-src/src/field_5x52_impl.h",
      "secp256k1-src/src/field_5x52_int128_impl.h",
      "secp256k1-src/src/field_5x52_asm_impl.h",
      "secp256k1-src/src/field.h",
      "secp256k1-src/src/field_impl.h",
      "secp256k1-src/src/group.h",
      "secp256k1-src/src/group_impl.h",
      "secp256k1-src/src/num.h",
      "secp256k1-src/src/num_impl.h",
      "secp256k1-src/src/num_gmp.h",
      "secp256k1-src/src/num_gmp_impl.h",
      "secp256k1-src/src/scalar_4x64.h",
      "secp256k1-src/src/scalar_4x64_impl.h",
      "secp256k1-src/src/scalar_8x32.h",
      "secp256k1-src/src/scalar_8x32_impl.h",
      "secp256k1-src/src/scalar.h",
      "secp256k1-src/src/scalar_impl.h",
      "secp256k1-src/src/util.h",
      "secp256k1-src/src/modules/schnorr/main_impl.h",
      "secp256k1-src/src/modules/schnorr/schnorr_impl.h",
      "secp256k1-src/src/modules/schnorr/schnorr.h",
      "secp256k1-src/src/modules/ecdh/main_impl.h",
      "secp256k1-src/src/modules/recovery/main_impl.h"
```

Edited "functions.cc" line 4-5 to:

``` c
#include "util.h"
#include "async.h"
```

Edited "util.cc" line 5 to:

``` c
#include "util.h"
```

Edited "util.h" line 7 and 8 to:

``` c
#include "secp256k1-src/include/secp256k1.h"
#include "secp256k1-src/include/secp256k1_recovery.h"
```

Edited "async.h" line 6 to:

``` c
#include "util.h"
```

Edited "secp256k1.cc" line 4-5 to:

``` c
#include "util.h"
#include "functions.h"
```

Got:

``` c
..\functions.cc(165): error C3861: 'secp256k1_ec_privkey_import': identifier not found [C:\Users\gunars\Documents\bitau
th\node_modules\secp256k1\build\secp256k1.vcxproj]
..\functions.cc(182): error C3861: 'secp256k1_ec_privkey_export': identifier not found [C:\Users\gunars\Documents\bitau
th\node_modules\secp256k1\build\secp256k1.vcxproj]
```

so I edited "functions.cc" corresponding lines to:

``` c
int results = 0;
```

Got bunch of warnings and:

``` c
LINK : fatal error LNK1181: cannot open input file 'C:\OpenSSL-Win32\lib\libeay32.lib' [C:\Users\gunars\Documents\bitau
th\node_modules\secp256k1\build\secp256k1.vcxproj]
```

so I downloaded OpenSSL from https://code.google.com/p/openssl-for-windows/downloads/detail?name=openssl-0.9.8k_WIN32.zip

And compiled successfully adding npm parameter --msvs_version=2015

Now how do I fix secp256k1_ec_privkey_import/secp256k1_ec_privkey_export error?

**Comment by wanderer** [comment]
@gunarssimkuns thanks for working on this! re: secp256k1_ec_privkey_import/secp256k1_ec_privkey_export idk but [stackoverflow](https://encrypted.google.com/search?hl=en&q=error%20C3861%3A%20identifier%20not%20found) suggests that it could be a problem with the header order. Maybe check that?

**Comment by simkuns** [comment]
Using search I found out that secp256k1_ec_privkey_import/secp256k1_ec_privkey_export functions are not defined anywhere in the sources, maybe I am missing something?

The only function I found was "secp256k1_ec_privkey_import_der()" in "secp256k1-src\contrib\lax_der_privatekey_parsing.h" file.

**Comment by wanderer** [comment]
@gunarssimkuns  It [here](https://github.com/bitcoin/secp256k1/blob/31994c8e5b8ec6693b2f8678b39d12d65c67067d/include/secp256k1.h#L449) and [here](https://github.com/bitcoin/secp256k1/blob/31994c8e5b8ec6693b2f8678b39d12d65c67067d/include/secp256k1.h#L427) Maybe you submodule is at a different commit?

**Comment by simkuns** [comment]
Yes, thank you now it compiles without errors.

I was using a different commit, so my "secp256k1/include/secp256k1.h" file looked a lot different.

**Comment by wanderer** [comment]
@gunarssimkuns if you got this to work, could you send in a PR with the changes you made?

**Comment by Maarten88** [comment]
Just setting with_gimp=false will make it compilable on Windows, with no other changes.
Download or compile OpenSSL in C:\OpenSSL-Win32 or OpenSSL-Win64, or add --openssl_root=[path] to point to an openssl installation somewhere else (it should have lib/libeay32.lib in it)

```
  "variables": {
    "conditions": [
      [
        "OS=='win'",
        {
            "with_gmp%": "false"
        },
        {
            "with_gmp%": "<!(utils/has_lib.sh gmpxx && utils/has_lib.sh gmp)"
        }
      ]
    ]
  },
```

**Comment by tcoulter** [comment]
@Maarten88 I tried following what you put above, but I'm still receiving errors on Windows 8.1 64bit. Mind taking a look? https://gist.github.com/anonymous/3416fe67808025e001e6

**Comment by tcoulter** [comment]
Here's what I had to do to get it to install on Windows: https://github.com/cryptocoinjs/secp256k1-node/pull/73

**Comment by fanatid** [comment]
@moshest @gunarssimkuns @Maarten88 please check version 3.0.1!

**Comment by moshest** [comment]
Thank you. I no longer use Windows :)

**Comment by axic** [comment]
@fanatid Shouldn't this be closed given #73 was merged?

**Comment by fanatid** [comment]
@axic yeah, I think we can close this issue.
